### PR TITLE
feat(runtime/http): websocket support

### LINKED ISSUE #10211: Support web sockets with hyper binding
Probably looks like this:
```js
let { respondWith, request } = await httpConn.nextRequest();
let { ws, response } = Deno.serveWebsocket(request);
respondWith(response);
// use web standard ws
```

Should use [tungstenite](https://crates.io/crates/tungstenite) 

Probably this should proceed #10210

### PR DESCRIPTION
**Pull request by crowlKats**
Closes #10211

### CODE REVIEW

File: runtime/js/40_http.js
Comment (ije) [inline runtime/js/40_http.js:72]: can we use `upgrade` instaed of `upgradeWebsocket ` ?

Comment (crowlKats) [reply runtime/js/40_http.js:72]: No, upgrading isn't something exclusive to websockets. that would end up being confusing.

**Review by ry — COMMENTED** [review]
Is this actually working? Would be very nice to have a test in cli/tests/unit/http_test.ts if so.

**Comment by crowlKats** [comment]
@ry well we first need #10365

File: runtime/js/40_http.js
Comment (bnoordhuis) [inline runtime/js/40_http.js:275]: Maybe add a `default:` that throws an illegal argument exception?

**Review by bnoordhuis — APPROVED** [review]
Needs regression tests but LGTM apart from that, thanks!

File: cli/dts/lib.deno.unstable.d.ts
Comment (nayeemrmn) [inline cli/dts/lib.deno.unstable.d.ts:1193]:
```suggestion
    upgradeWebSocket(): Promise<WebSocketUpgrade>;
```

File: runtime/js/40_http.js
Comment (nayeemrmn) [reply runtime/js/40_http.js:72]: > Of course that would make sense if we would support upgrading other kind of conns out of the box

Other hypothetical upgrades would have different return types, so they should just have individual names anyway.

**Comment by ry** [comment]
@crowlKats I'm a bit confused about how this depends on #10365... I haven't dived into this PR but I'm not sure why we would need to expose WebSocketStream in order to hook up the hyper-tungstenite bindings?

I don't care so much about WebSocketStream support - since it's still an tentative API - but I'm rather keen to land this PR.

File: runtime/js/40_http.js
Comment (nayeemrmn) [inline runtime/js/40_http.js:223]:
```suggestion
  function createUpgradeWebSocket(rid) {
```

File: runtime/js/40_http.js
Comment (lucacasonato) [inline runtime/js/40_http.js:72]: instead of returning `upgradeWebsocket` here there should be a global `Deno.upgradeWebSocket` that takes a `Request` and returns both a `WebSocket(Stream)` and a `Response`. The connection will only be upgraded after the user passes this `Response` to `respondWith`.

File: runtime/js/40_http.js
Comment (lucacasonato) [inline runtime/js/40_http.js:318]: header names should all be lower-case in the header list

**Comment by lucacasonato** [comment]
Blocked on https://github.com/hyperium/hyper/pull/2587

**Review by lucacasonato — APPROVED** [review]
LGTM. Thanks for this great feature!

We need to add some documentation for this to https://deno.land/manual@v1.11.5/runtime/http_server_apis in a follow up commit.

**Review by lucacasonato — CHANGES_REQUESTED** [review]
Oh actually `upgradeWebsocket` is still async. That needs to become synchronous.

**Review by lucacasonato — APPROVED** [review]
LGTM now. Thanks!

**Pull request merged**
